src/plugins/arithmetic.ts

```diff
@@ -283,3 +284,126 @@ export function mulOrDivToBitShift(fromMul = true, fromDiv = true): Plugin {
 }
 
 export const bitShiftPlugins = [bitShiftToMulOrDiv(), mulOrDivToBitShift()];
+
+export type IntDecomposition = [
+  // [k,b,e,d,cost] represents a value k * pow(b,e) + d, cost = lg(k)+lg(b)+lg(d)
+  bigint,
+  bigint,
+  bigint,
+  bigint,
+  number
+];
+
+// assert (1000 ≤ |n|)
+export function decomposeInt(n: bigint): IntDecomposition[] {
+  return decomposeAnyInt(n, n);
+}
+
+// assert (1000 ≤ x ≤ y || x ≤ y ≤ -1000)
+export function decomposeAnyInt(x: bigint, y: bigint): IntDecomposition[] {
+  return x > 0
+    ? _decomposeAnyInt(x, y)
+    : _decomposeAnyInt(-y, -x).map(([k, b, e, d, c]) => [-k, b, e, -d, c]);
+}
+
+function lg(n: bigint): number {
+  if (n < 0) n = -n;
+  return n > 1000000000000
+    ? 2 + n.toString(16).length
+    : n > 99
+    ? n.toString().length
+    : n > 9
+    ? 2
+    : 1;
+}
+
+function betterOrEqual(
+  [k1, b1, e1, d1, c1]: IntDecomposition,
+  [k2, b2, e2, d2, c2]: IntDecomposition
+): boolean {
+  return (
+    (b1 === 2n || b2 !== 2n) &&
+    (b1 === 10n || b2 !== 10n) &&
+    (k1 === 1n || k2 !== 1n) &&
+    (d1 === 0n || d2 !== 0n) &&
+    c1 <= c2
```

Comment:
How about a little heuristic that limits the power of the flags to outweigh the length?
```suggestion
    c1 + 10 < c2 ||
    ((b1 === 2n || b2 !== 2n) &&
      (b1 === 10n || b2 !== 10n) &&
      (k1 === 1n || k2 !== 1n) &&
      (d1 === 0n || d2 !== 0n) &&
      c1 <= c2)
```

Reply:
What are you trying to fix? Do you have an example?

I think that the `b=2` and `b=10` flags shouldn't be limited because some languages don't have a power function.
I'm not sure about the other two flags. They can be expensive: `*k` in Java is `.multiply(BigInteger.valueOf(k))`. But I don't have another example.

The `decompose(Any)?Int` functions are language agnostic. But we can add filters to the `decomposeIntLiteral` plugin since we can pass language specific info like `hasPow`, `hasBitShift`, or `bigIntegerOperatorsAreAbsurdlyLong`.